xmms-shell/xmms-shell.1
.\"                                      Hey, EMACS: -*- nroff -*-
.\" First parameter, NAME, should be all caps
.\" Second parameter, SECTION, should be 1-8, maybe w/ subsection
.\" other parameters are allowed: see man(7), man(1)
.TH XMMS-SHELL SECTION "February  8, 2002"
.\" Please adjust this date whenever revising the manpage.
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7)
.SH NAME
xmms-shell \- Command-line utility for controlling XMMS externally
.SH SYNOPSIS
.B xmms-shell [OPTIONS]
.SH OPTIONS
.TP
.B \-e [expr], \-\-eval [expr]
Evaluate expr and exit
.TP
.B \-h, \-\-help
Display this message and exit
.TP
.B \-n [s], \-\-session [s]
Specify session ID
Show version of program.
.SH SEE ALSO
.BR xmms (1)
.SH AUTHOR
XMMS-Shell and this manual page were written by Logan Hanks <logan@vt.edu>.



xmms-shell/ChangeLog
Sat Oct 19 19:49:09 UTC 2002  Logan Hanks  <logan@vt.edu>

  * Added in long-awaited changes to make xmms-shell compile under g++ 3
    (patches from several sources)
  * Titles of tracks displayed by LIST are now aligned, and the current
    track is now marked (request #514146)
  * RANDOMTRACK renamed to RANDOM-TRACK (alias to RANDOMTRACK also added)
  * Added CURRENT-TRACK command (request #514146)
  * Added REMOVE command (request #514146)

Thu Oct 17 19:30:49 UTC 2002  Logan Hanks  <logan@vt.edu>

  * Released as 0.99.1
  * WINDOW SHOW now has the correct behavior (closes bug #583730)
  * FADE now advances to next track (closes Debian bug #120161)

Tue Feb 12 01:30:18 UTC 2002  Logan Hanks  <logan@vt.edu>

  * Fixed a typo that caused downvolume to be the same as upvolume (copy and
    paste is a sign of poor design!)
  * The quit field of the CommandContext object is now explicitly
    initialize to false (why didn't g++ -Wall catch this?)
  * Implemented XMMSQUIT command (idea courtesy of Roderick Schertler)
  * include/playlist.h is now mentioned in include/Makefile.am
  * Implemented RANDOMTRACK command, with only a minor attempt at making it
    seem random.  Rather than hard-coding some behavior for the NEXT and
    PREVIOUS commands when shuffle-mode is on, RAND could be used instead.
  * Went ahead and hard-coded some behavior for the NEXT and PREVIOUS
    commands when shuffle-mode is on.  Note that this only works on versions
    of XMMS that provide xmms_remote_is_shuffle (I forget which version
    introduced this).  XMMS's support of this seems really poor, so maybe
    RANDOMTRACK is better.

Mon Feb 11 03:46:06 UTC 2002  Logan Hanks  <logan@vt.edu>

  * Released as 0.99.0
  * Massively rewritten
    - Rewrote how commands are specified and how they provide help text
    - Implemented simple quoting in the command parser
    - If readline is enabled, filenames will be properly quoted
    - In the future, more complex scripting (rather than a simple batch
      mode) may be implemented, so command system redesigned to reflect this
    - All commands should now give some sort of feedback
    - All of the functionality of 0.2.2 should be emulated


Thu May 18 02:46:59 UTC 2000  Logan Hanks  <logan@vt.edu>

  * Released as 0.2.2
  * Moved old changes to ChangeLog.old, switched to new format
  * Made more improvements to the installation process
  * Added resetdevice command
  * Added preamp and band commands (only available with a very
    recent copy of XMMS checked out from CVS)



xmms-shell/src/xmms-shell.cc
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <getopt.h>
#include <xmmsctrl.h>
#include "config.h"
#include "command.h"
#include "eval.h"
#include "general.h"
#include "getline.h"
#include "misc.h"
#include "playback.h"
#include "playlist.h"
#include "volume.h"
#include "window.h"

static struct option long_options[] = {
	{ "session", 1, 0, 'n' },
	{ "eval", 1, 0, 'e' },
	{ "help", 0, 0, 'h' },
	{ 0, 0, 0, 0 }
};

static char *program_name;

static void display_usage(FILE *f)
{
	fprintf(f, "\n");
	fprintf(f, "Usage: %s [OPTIONS]\n", program_name);
	fprintf(f, "\n");
	fprintf(f, "Options:\n");
	fprintf(f, "\n");
	fprintf(f, "  -e [expr], --eval [expr] Evaluate expr and exit\n");
	fprintf(f, "  -h, --help               Display this message and exit\n");
	fprintf(f, "  -n [s], --session [s]    Specify session ID\n");
	fprintf(f, "\n");
	fprintf(f, "If no expression is specified on the command line the program will\n");
	fprintf(f, "switch into interactive mode.\n");
	fprintf(f, "\n");
}

static int eval_loop(int session_id, FILE *in)
{
	int quit = 0, retval = 127;
    ScriptContext *context;

    if(isatty(fileno(in))) {
        context = new InteractiveContext();
    } else {
        context = new FileContext(in);
    }
    context->set_session(Session(session_id));
	//while(!quit && (line = getline("xmms-shell> ")))
    try {
        while(!quit) {
            string line = context->get_line();

		    retval = eval_command_string(context, line, quit, TRUE);
        }
    } catch(EOFException ex) {
    }
	if(!quit) {
		printf("\n");
    }
	return retval;
}

int main(int argc, char **argv)
{
	int opt, optind;
	int session_id = 0;
	char *do_expr = NULL;

	program_name = argv[0];
	while((opt = getopt_long(argc, argv, "n:e:h", long_options, &optind)) != -1) {
		switch(opt) {
			case 'h':
			case 0:
			case ':':
			case '?':
				display_usage(stderr);
				return 0;
			case 'n':
				session_id = atoi(optarg);
				break;
			case 'e':
				do_expr = optarg;
				break;
		}
	}
	if(!xmms_remote_is_running(session_id)) {
		fprintf(stderr, "XMMS is not running under the session identifier ``%d''\n", session_id);
		return 1;
	}

	general_init();
	getline_init();
	misc_init();
	playback_init();
	playlist_init();
	volume_init();
	window_init();

	command_init();

	if(do_expr) {
        ScriptContext *context = new StringContext(do_expr);
        string line = context->get_line();
        int quit;

		return eval_command_string(context, line, quit, FALSE);
    }
	return eval_loop(session_id, stdin);
}

